### PR TITLE
feat: add warning for binary string in "application/json" or parameter

### PR DESCRIPTION
**Pull request by barrett-schonefeld**
Changes:
- added a findOctetSequencePaths function to handle the logic of finding schema type: string, format: binary for cases of nested arrays, objects, nested arrays of type object, objects with properties that are nested arrays, and objects with properties that are objects, and the simple case where a schema uses type: string, format: binary directly. This function takes a resolved schema and returns a list of paths to octet sequences (empty list if none found). The function accepts the path both as an array and a string and returns the path in the same format received.
- added logic to handle application/json request bodies that use schema type: string, format: binary
- added logic to handle application/json response bodies of type: string, format: binary
- added logic to handle parameters of type: string, format: binary
- removed 'binary' as a valid format for type: string parameters. parameters of type: string, format: binary will result in "type+format not well-defined" error
- added "json_or_param_binary_string" as to .defaultsForValidator as a warning in the shared.schemas section
- added "json_or_param_binary_string" configuration option to the README.md rules and defaults sections in the schemas section
- changed findOctetSequencePaths to accept the path both as a string and an array.

Refactoring in responses.js:
- moved the check for 204-response with content to be part of the other success code checks
- added a function to get all response codes, all status codes, and all success codes

Tests:
- added tests to ensure warnings are issued for request bodies, response bodies, and parameters with schema, type: string, format: binary
- added complex tests to exercise combinations of nested arrays and objects that contain schema type: string, format: binary (complex tests done on response bodies)

### CODE REVIEW

File: src/plugins/validation/oas3/semantic-validators/responses.js [hunk 1]
```diff
@@ -84,7 +93,49 @@ module.exports.validate = function({ resolvedSpec }, config) {
...
+          const octetSequencePaths = octetSequences(
+            responseBodyContent[mimeType].schema,
+            schemaPath
+          );
+          const message =
+            'No well-defined way to represent bare octets (type: string, format: binary) as JSON.';
+          for (const p of octetSequencePaths) {
+            result[binaryStringStatus].push({
+              path: p.split('.'),
+              message: message
```
Comment (mkistler) [inline src/plugins/validation/oas3/semantic-validators/responses.js]: There seems to be contention in how to represent the path here.  We need to join it to call `octetSequencePaths`, but then split it again to set result.  Can't we just use one format consistently?

File: test/plugins/validation/oas3/operations.js [hunk 2]
```diff
@@ -191,6 +250,16 @@ describe('validation plugin - semantic - operations - oas3', function() {
...
+    const config = {
+      operations: {
+        no_request_body_content: 'error',
+        no_request_body_name: 'warning'
+      },
+      schemas: {
+        binary_string: 'warning'
+      }
+    };
+
```
Comment (mkistler) [inline test/plugins/validation/oas3/operations.js]: It seems odd that we would have to add this config for existing tests.  Why is this?

File: test/plugins/validation/oas3/parameters.js [hunk 3]
```diff
@@ -143,6 +179,18 @@ describe('validation plugin - semantic - parameters - oas3', function() {
...
+        no_in_property: 'error',
+        invalid_in_property: 'error',
+        missing_schema_or_content: 'error',
+        has_schema_and_content: 'error'
+      },
+      schemas: {
+        binary_string: 'warning'
+      }
+    };
+
```
Comment (mkistler) [inline test/plugins/validation/oas3/parameters.js]: Again ... why are these needed when they weren't needed previously?

File: src/plugins/validation/oas3/semantic-validators/responses.js (hunk 1 above)
Comment (barrett-schonefeld) [reply src/plugins/validation/oas3/semantic-validators/responses.js]: Some of the semantic validators use arrays and others use strings with `.` as the delimiter. For this PR I just matched the convention of each validator, but I would like to standardize the paths to use arrays in the next PR.

File: src/plugins/validation/oas3/semantic-validators/responses.js [hunk 4]
```diff
@@ -84,7 +93,49 @@ module.exports.validate = function({ resolvedSpec }, config) {
...
+}
+
+function isStatusCode(code) {
   const allowedFirstDigits = ['1', '2', '3', '4', '5'];
   return code.length === 3 && allowedFirstDigits.includes(code.slice(0, 1));
 }
+
+function validateResponseBody(
+  responseObj,
+  allResponseCodes,
```
[the remark this answers is not in the document]
Comment (barrett-schonefeld) [reply src/plugins/validation/oas3/semantic-validators/responses.js]: Yes it is.

File: test/plugins/validation/oas3/operations.js (hunk 2 above)
Comment (barrett-schonefeld) [reply test/plugins/validation/oas3/operations.js]: Since this binary string rule spans multiple semantic validators, we have to get the rule from the shared schema rules with `const configSchemas = config.schemas`.

The parameters, responses, and operations validators previously only used the OAS3-specific rules, but the way it is implemented now requires the shared schema rules to be present. Later it is used:

`const binaryStringStatus = configSchemas.binary_string;
  if (binaryStringStatus !== 'off') {`

I could wrap this with:

`if (configSchemas) {
    const binaryStringStatus = configSchemas.binary_string;`

File: test/plugins/validation/oas3/parameters.js (hunk 3 above)
Comment (barrett-schonefeld) [reply test/plugins/validation/oas3/parameters.js]: (See above)

File: src/plugins/validation/oas3/semantic-validators/parameters.js [hunk 5]
```diff
@@ -66,8 +73,39 @@ module.exports.validate = function({ jsSpec }, config) {
...
+            'Parameters should not contain binary (type: string, format: binary) values.';
+          result[binaryStringStatus].push({
+            path: p.split('.'),
+            message: message
+          });
+        }
+      }
     }
   });
-
```
Comment (dpopp07) [inline src/plugins/validation/oas3/semantic-validators/parameters.js]: No need to remove this whitespace

File: test/plugins/validation/oas3/operations.js [hunk 6]
```diff
@@ -2,10 +2,19 @@ const expect = require('expect');
 const {
   validate
 } = require('../../../../src/plugins/validation/oas3/semantic-validators/operations');
-const config = require('../../../../src/.defaultsForValidator').defaults.oas3;
```
Comment (dpopp07) [inline test/plugins/validation/oas3/operations.js]: Wait.. why have you removed this line? This is meant to handle the config so that you don't have to define it for every independent test. It should work for your use case

File: test/plugins/validation/oas3/operations.js (hunk 2 above)
Comment (dpopp07) [reply test/plugins/validation/oas3/operations.js]: I now see that this answers my comment above. I think we might want to re-consider the approach here. Possibly define separate rules for each section to give more fine-grained control. We can discuss this with @mkistler

File: src/plugins/validation/oas3/semantic-validators/operations.js [hunk 7]
```diff
@@ -78,10 +84,32 @@ module.exports.validate = function({ resolvedSpec, jsSpec }, config) {
...
+                const octetSequencePaths = octetSequences(
+                  requestBodyContent[mimeType].schema,
+                  schemaPath
+                );
+                const message =
+                  'No well-defined way to represent bare octets (type: string, format: binary) as JSON.';
+                for (const p of octetSequencePaths) {
+                  result[binaryStringStatus].push({
+                    path: p,
+                    message: message
```
[the remark this answers is not in the document]
Comment (dpopp07) [reply src/plugins/validation/oas3/semantic-validators/operations.js]: I only commented on a few, but all of these instances need to be fixed

File: src/plugins/validation/oas3/semantic-validators/operations.js [hunk 8]
```diff
@@ -78,10 +84,32 @@ module.exports.validate = function({ resolvedSpec, jsSpec }, config) {
                 'Operations with non-form request bodies should set a name with the x-codegen-request-body-name annotation.';
               result[checkStatus].push({
                 path: `paths.${pathName}.${opName}`,
-                message
+                message: message
```
[the remark this answers is not in the document]
Comment (barrett-schonefeld) [reply src/plugins/validation/oas3/semantic-validators/operations.js]: Other logging in the module defines the key, so I added for consistency.

[the remark this answers is not in the document]
Comment (dpopp07) [reply src/plugins/validation/oas3/semantic-validators/operations.js]: I see. In that case, I would recommend you change the rest of them to match what this was before

File: test/plugins/validation/oas3/parameters.js [hunk 9]
```diff
@@ -224,7 +284,219 @@ describe('validation plugin - semantic - parameters - oas3', function() {
...
+      schemas: {
+        json_or_param_binary_string: 'warning'
+      }
+    };
+
+    const spec = {
+      components: {
+        parameters: {
+          GoodParam: {
+            schema: {}
```
Comment (dpopp07) [inline test/plugins/validation/oas3/parameters.js]: I would actually recommend making this test about the schema field not even being there. That's often a good thing to verify in JS code. Just make `GoodParam` an empty object.

File: src/.defaultsForValidator.js [hunk 10]
```diff
@@ -64,7 +64,8 @@ const defaults = {
       'description_mentions_json': 'warning',
       'array_of_arrays': 'warning',
       'property_case_convention': [ 'error', 'lower_snake_case'],
-      'enum_case_convention': [ 'error', 'lower_snake_case']
+      'enum_case_convention': [ 'error', 'lower_snake_case'],
+      'binary_string': 'warning'
```
Comment (dpopp07) [inline src/.defaultsForValidator.js]: Don't forget - this needs to match what is in the README/code

File: test/plugins/validation/oas3/operations.js (hunk 2 above)
Comment (barrett-schonefeld) [reply test/plugins/validation/oas3/operations.js]: Added a 'schemas' section to the OAS3 rules and put the json_or_param_binary_string config option there.

Then removed the individual configuration objects in each test and replaced it by importing the default OAS3 rules.

File: test/plugins/validation/oas3/parameters.js [hunk 11]
```diff
@@ -1,4 +1,5 @@
 const expect = require('expect');
+
```
Comment (dpopp07) [inline test/plugins/validation/oas3/parameters.js]: Don't need to add this whitespace
